### PR TITLE
Standardize API/DB Errors

### PR DESCRIPTION
**Pull request by akinsey**

### CODE REVIEW

File: modules/ept-ads/db/rounds/rotate.js [hunk 1]
```diff
@@ -13,10 +14,10 @@ module.exports = function(round) {
     return client.queryAsync(check, [round])
     .then(function(response) {
       if (response.rows.length < 1) {
-        throw new RoundNotFoundError('This round does not exists');
+        throw new NotFoundError('This round does not exists');
       }
       else if (response.rows[0].start_time) {
-        throw new AlreadyUsedError('This round has already been in use');
+        throw new ConflictError('This round has already been in use');
```
Comment (unenglishable) [inline modules/ept-ads/db/rounds/rotate.js]: 'This round is already in use'

File: modules/ept-ads/db/rounds/rotate.js [hunk 2]
```diff
@@ -13,10 +14,10 @@ module.exports = function(round) {
     return client.queryAsync(check, [round])
     .then(function(response) {
       if (response.rows.length < 1) {
-        throw new RoundNotFoundError('This round does not exists');
+        throw new NotFoundError('This round does not exists');
```
Comment (unenglishable) [inline modules/ept-ads/db/rounds/rotate.js]: 'This round does not exist'

File: modules/ept-ads/db/rounds/rotate.js (hunk 1 above)
Comment (unenglishable) [inline modules/ept-ads/db/rounds/rotate.js]: or 'This round has already been used'

File: modules/ept-ads/db/rounds/rotate.js (hunk 2 above)
Comment (akinsey) [inline modules/ept-ads/db/rounds/rotate.js]: Done boss

File: modules/ept-ads/db/rounds/rotate.js (hunk 1 above)
Comment (akinsey) [inline modules/ept-ads/db/rounds/rotate.js]: Done boss

File: modules/ept-ads/routes/factoids/enable.js [hunk 3]
```diff
@@ -28,10 +42,12 @@ module.exports = {
...
 
     if (factoidId === 'all') {
-      promise = db.factoids.enableAll();
+      promise = db.factoids.enableAll()
+      .error(request.errorMap.toHttpError);
     }
     else {
-      promise = db.factoids.enable(factoidId);
+      promise = db.factoids.enable(factoidId)
+      .error(request.errorMap.toHttpError);
```
Comment (unenglishable) [inline modules/ept-ads/routes/factoids/enable.js]: DRY this by moving error outside of if/else block

File: modules/ept-ads/text/info.txt [hunk 4]
```diff
@@ -1,3 +1,3 @@
 Forum ad slots are auctioned off weekly.
 <a href="/ads/analytics">Here are the ad stats. </a>
-See the curernt auction thread for more info.
+See the current auction thread for more info.
```
Comment (unenglishable) [inline modules/ept-ads/text/info.txt]: nice 👍

File: modules/ept-auto-moderation/db/addRule.js [hunk 5]
```diff
@@ -24,7 +26,7 @@ module.exports = function(rule) {
       rule.id = rows[0].id;
       return rule;
     }
-    else { throw Error('Could Not Create Rule'); }
+    else { throw new CreationError('Could Not Create Rule'); }
```
Comment (unenglishable) [inline modules/ept-auto-moderation/db/addRule.js]: Why is this in title case?  All the others are in sentence case.

File: modules/ept-boards/db/getBoardWriteAccess.js [hunk 6]
```diff
@@ -12,7 +13,7 @@ module.exports = function(boardId, userPriority) {
   return db.sqlQuery(q, [boardId])
   .then(function(rows) {
     if (rows.length > 0 ) { return rows[0].postable_by; }
-    else { throw new NotFoundError(); }
+    else { throw new NotFoundError('Field \'postable_by\' Not Found'); }
```
Comment (unenglishable) [inline modules/ept-boards/db/getBoardWriteAccess.js]: Maybe say something like: 'No board write access found for board.'

File: modules/ept-ignore-users/routes/ignore.js [hunk 7]
```diff
@@ -16,7 +30,7 @@ module.exports = {
     .then(function() { return { userId: ignoreUserId, ignored: true }; })
     .error(function(err) {
       if (err.constraint) { return {}; }
-      throw err;
```
Comment (unenglishable) [inline modules/ept-ignore-users/routes/ignore.js]: sick. 😷

File: modules/ept-messages/routes/findUser.js [hunk 8]
```diff
@@ -3,14 +3,18 @@ var Joi = require('joi');
 /**
   * @apiVersion 0.4.0
   * @apiGroup Messages
-  * @api {GET} /messages/users/{username} Get ID for username
+  * @api {GET} /messages/users/{username} Message Receiver Lookup
   * @apiName FindUserMessages
   * @apiPermission User
-  * @apiDescription Get the id for the given username
+  * @apiDescription Querry possible username matches and returns their ids for use in message delivery
```
Comment (unenglishable) [inline modules/ept-messages/routes/findUser.js]: *Query

File: modules/ept-posts/db/getBoardWriteAccess.js [hunk 9]
```diff
@@ -17,7 +17,7 @@ module.exports = function(postId, userPriority) {
   return db.sqlQuery(q, [postId])
   .then(function(rows) {
     if (rows.length > 0 ) { return rows[0].postable_by; }
-    else { throw new NotFoundError(); }
+    else { throw new NotFoundError('Field \'postable_by\' Not Found'); }
```
Comment (unenglishable) [inline modules/ept-posts/db/getBoardWriteAccess.js]: another one, as explained before

File: modules/ept-posts/db/getPostsBoardInBoardMapping.js [hunk 10]
```diff
@@ -1,17 +1,17 @@
...
+var NotFoundError = errors.NotFoundError;
 
 module.exports = function(postId, userPriority) {
   postId = helper.deslugify(postId);
   var q = 'SELECT t.board_id FROM posts p LEFT JOIN threads t ON p.thread_id = t.id WHERE p.id = $1';
   return db.sqlQuery(q, [postId])
   .then(function(rows) {
     if (rows.length > 0 ) { return rows[0].board_id; }
-    else { throw new NotFoundError(); }
+    else { throw new NotFoundError('Field \'board_id\' Not Found'); }
```
Comment (unenglishable) [inline modules/ept-posts/db/getPostsBoardInBoardMapping.js]: Another one, as explained before

File: modules/ept-posts/routes/pageByUser.js [hunk 11]
```diff
@@ -13,13 +13,31 @@ var common = require(path.normalize(__dirname + '/../common'));
...
+  * @apiSuccess {timestamp} posts.updated_at The updated at timestamp of the post
+  * @apiSuccess {timestamp} posts.imported_at The imported at timestamp of the post
+  * @apiSuccess {string} posts.board_id The id of the board containing the post
+  * @apiSuccess {string} posts.thread_title The title of the thread the post is in
+  * @apiSuccess {string} posts.avatar The avatar of the user who made the post
+  * @apiSuccess {object} posts.user Object containing user data about the author of the post
+  * @apiSuccess {string} posts.user.id The id of the user
   *
-  * @apiError (Error 500) InternalServerError There was an issue finding posts for the user
+  * @apiError (Error 500) InternalServerError There was an issue paging posts for user
```
Comment (unenglishable) [inline modules/ept-posts/routes/pageByUser.js]: nice 👍

File: modules/ept-threads/routes/lockPoll.js [hunk 12]
```diff
@@ -5,18 +5,18 @@ var Joi = require('joi');
   * @apiGroup Threads
   * @api {POST} /threads/:threadId/polls/:pollId/lock Lock Poll
   * @apiName LockPoll
-  * @apiPermission Super Administrator, Administrator, Global Moderator, Moderator, User
+  * @apiPermission Super Administrator, Administrator, Global Moderator, Moderator, User (Poll creator only)
```
Comment (unenglishable) [inline modules/ept-threads/routes/lockPoll.js]: 👍

File: modules/ept-threads/routes/purge.js [hunk 13]
```diff
@@ -10,9 +10,9 @@ var Joi = require('joi');
   *
   * @apiParam {string} id The unique id of the thread to purge
   *
-  * @apiUse ThreadObjectSuccess2
```
Comment (unenglishable) [inline modules/ept-threads/routes/purge.js]: 👎

File: modules/ept-users/routes/reactivate.js [hunk 14]
```diff
@@ -30,7 +30,10 @@ module.exports = {
   },
   handler: function(request, reply) {
     var userId = request.params.id;
-    var promise = request.db.users.reactivate(userId);
+    var promise = request.db.users.reactivate(userId)
+    .then(function() { return {}; })
```
Comment (unenglishable) [inline modules/ept-users/routes/reactivate.js]: we need to figure out how to do this correctly.  will post an issue about it.

File: modules/ept-users/routes/removeInvitation.js [hunk 15]
```diff
@@ -11,7 +11,7 @@ var Joi = require('joi');
   *
   * @apiSuccess {string} message Invitation removal success message
   *
-  * @apiError BadRequest There was an error removing the invitation
+  * @apiError (Error 500) InternalServerError There was error getting removing the invite
```
Comment (unenglishable) [inline modules/ept-users/routes/removeInvitation.js]: Is it getting or removing?  Or is it really both?

File: modules/ept-users/routes/removeInvitation.js [hunk 16]
```diff
@@ -27,7 +27,9 @@ module.exports = {
     // remove invitation
     var email = request.payload.email;
     var promise = request.db.users.removeInvite(email)
-    .then(function() { return { message: 'Invitation Removed.' }; });
+    .then(function() { return { message: 'Invitation Removed.' }; })
```
Comment (unenglishable) [inline modules/ept-users/routes/removeInvitation.js]: Should all success instances return a similar message?  or should they just return nothing?

Will discuss.

File: modules/ept-watchlist/routes/pageThreads.js [hunk 17]
```diff
@@ -7,12 +7,20 @@ var Joi = require('joi');
...
+  * @apiSuccess {boolean} hasMoreThreads Boolean indicating if there are more pages of threads
+  * @apiSuccess {object[]} threads An array containing watched thread data
+  * @apiSuccess {string} threads.id The unique id of the watched thread
+  * @apiSuccess {number} threads.post_count The post count of the watched thread
+  * @apiSuccess {number} threads.view_count The view count of the watched thread
+  * @apiSuccess {string} threads.board_name The name of the board the thread is in
+  * @apiSuccess {string} threads.title The title of the thread being watched
   *
-  * @apiError (Error 500) InternalServerError There was an issue retrieving the watchlist threads.
+  * @apiError (Error 500) InternalServerError There was an issue retrieving the thread watchlist threads.
```
Comment (unenglishable) [inline modules/ept-watchlist/routes/pageThreads.js]: ???

File: package.json [hunk 18]
```diff
@@ -40,7 +40,7 @@
     "del": "^2.2.0",
     "dot": "^1.0.3",
     "dotenv": "^2.0.0",
-    "epochtalk-core-pg": "epochtalk/core-pg",
+    "epochtalk-core-pg": "epochtalk/core-pg#standardization-db",
```
Comment (unenglishable) [inline package.json]: revert before merging

File: server/routes/admin/moderators/config.js [hunk 19]
```diff
@@ -8,10 +8,13 @@ var Joi = require('joi');
   * @apiPermission Super Administrator, Administrator,
   * @apiDescription Add a moderator to a board.
   *
-  * @apiParam (Payload) {string[]} usernames Array of ids of the user to add as a moderator.
```
Comment (unenglishable) [inline server/routes/admin/moderators/config.js]: LOL 😛

File: server/routes/admin/settings/config.js [hunk 20]
```diff
@@ -15,6 +15,8 @@ var defaultVarsPath = path.normalize(__dirname + varsDir + '/_default-variables.
 var customPath = '/../../../../content/sass/_custom-variables.scss';
 var customVarsPath = path.normalize(__dirname + customPath);
 
+var ConfigError = Promise.OperationalError;
```
Comment (unenglishable) [inline server/routes/admin/settings/config.js]: Should we define this in the errors file?

File: server/routes/admin/settings/config.js [hunk 21]
```diff
@@ -705,7 +784,7 @@ function sanitizeConfigs(configurations, saasMode) {
 
 function checkEmailerConfig(emailer) {
   return new Promise(function(resolve, reject) {
-    if (!emailer) { return reject(new Error('Emailer configuration not found.')); }
+    if (!emailer) { return reject(new ConfigError('Emailer configuration not found.')); }
```
Comment (unenglishable) [inline server/routes/admin/settings/config.js]: Figure all of these out.

File: server/routes/admin/settings/config.js [hunk 22]
```diff
@@ -714,14 +793,14 @@ function checkEmailerConfig(emailer) {
     if (!emailer.user) { errors.push('Emailer User not found.'); }
     if (!emailer.pass) { errors.push('Emailer Password not found.'); }
 
-    if (errors.length > 0) { return reject(new Error(errors.join('\n'))); }
+    if (errors.length > 0) { return reject(new ConfigError(errors.join('\n'))); }
```
Comment (unenglishable) [inline server/routes/admin/settings/config.js]: used here

File: server/routes/admin/settings/config.js [hunk 23]
```diff
@@ -714,14 +793,14 @@ function checkEmailerConfig(emailer) {
...
     else { return resolve(); }
   });
 }
 
 function checkImagesConfig(images) {
   return new Promise(function(resolve, reject) {
-    if (!images) { return reject(new Error('Images configuration not found')); }
+    if (!images) { return reject(new ConfigError('Images configuration not found')); }
```
Comment (unenglishable) [inline server/routes/admin/settings/config.js]: and here

File: server/routes/admin/settings/config.js [hunk 24]
```diff
@@ -750,7 +829,7 @@ function checkImagesConfig(images) {
       if (!images.s3.secretKey) { errors.push('S3 Secret Key not set.'); }
     }
 
-    if (errors.length > 0) { return reject(new Error(errors.join('\n'))); }
+    if (errors.length > 0) { return reject(new ConfigError(errors.join('\n'))); }
```
Comment (unenglishable) [inline server/routes/admin/settings/config.js]: also here

File: modules/ept-ads/routes/factoids/enable.js (hunk 3 above)
Comment (unenglishable) [inline modules/ept-ads/routes/factoids/enable.js]: this is okay.

**Comment by unenglishable** [comment]
it's okay.  not great, but pretty good.

**Pull request merged**